File: Sources/BioToolsApp/AuthModels.swift
import Foundation

struct UserProfile: Codable, Equatable, Sendable {
    let id: String
    let email: String
    let realName: String
    let labRole: String
    let status: String
    let reviewReason: String?
    let authorizationExpiresAt: Int64?
    let authorizationPermanent: Bool?
}

struct DeviceProfile: Codable, Identifiable, Equatable {
    let id: String
    let platform: String
    let deviceName: String
    let appVersion: String
    let firstSeenAt: Int64
    let lastSeenAt: Int64
    let revokedAt: Int64?
    let current: Bool
}

struct AuthTokens: Codable, Equatable, Sendable {
    let accessToken: String
    let accessExpiresAt: Int64
    let refreshToken: String
    let refreshExpiresAt: Int64
    let offlineLicense: String
    let offlineLicenseExpiresAt: Int64
    let serverTime: Int64
}

struct StoredSession: Codable, Equatable, Sendable {
    let user: UserProfile
    let tokens: AuthTokens
}

struct SavedLoginCredentials: Codable, Equatable, Sendable {
    let email: String
    let password: String
}

struct BackendAuthorization: Equatable {
    let offlineLicense: String
    let installationHash: String
    let publicJWK: String
    let omicsKeyB64: String
}

enum AuthPhase: Equatable {
    case checking
    case signedOut
    case unverified(String)
    case pending(String)
    case rejected(String)
    case suspended(String)
    case authorized(UserProfile, expiresAt: Date, offline: Bool)
    case configurationMissing(String)
}

struct OfflineLicenseClaims: Codable {
    let typ: String
    let sub: String
    let device: String
    let iat: Int64
    let exp: Int64
    let version: Int
    let omicsKeyB64: String

    enum CodingKeys: String, CodingKey {
        case typ, sub, device, iat, exp, version
        case omicsKeyB64 = "omics_key_b64"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        typ = try container.decode(String.self, forKey: .typ)
        sub = try container.decode(String.self, forKey: .sub)
        device = try container.decode(String.self, forKey: .device)
        iat = try container.decode(Int64.self, forKey: .iat)
        exp = try container.decode(Int64.self, forKey: .exp)
        version = try container.decode(Int.self, forKey: .version)
        omicsKeyB64 = try container.decodeIfPresent(String.self, forKey: .omicsKeyB64) ?? ""
    }
}

enum AuthClientError: LocalizedError {
    case server(code: String, message: String, status: Int)
    case invalidResponse
    case network(Error)
    case configuration(String)

    var errorDescription: String? {
        switch self {
        case let .server(_, message, _): message
        case .invalidResponse: "授权服务返回了无效数据。"
        case let .network(error): "无法连接授权服务：\(error.localizedDescription)"
        case let .configuration(message): message
        }
    }

    var isExplicitRevocation: Bool {
        guard case let .server(code, _, status) = self else { return false }
        return status == 401 || status == 403 || [
            "AUTHORIZATION_REVOKED", "AUTHORIZATION_EXPIRED", "ACCOUNT_SUSPENDED", "ACCOUNT_DELETED", "SESSION_EXPIRED"
        ].contains(code)
    }
}

File: windows/MyBioTools.Windows/AuthModels.cs
using System.Text.Json.Serialization;

namespace MyBioTools.Windows;

internal sealed record UserProfile(
    string Id,
    string Email,
    string RealName,
    string LabRole,
    string Status,
    string? ReviewReason,
    long? AuthorizationExpiresAt,
    bool? AuthorizationPermanent);

internal sealed record DeviceProfile(
    string Id,
    string Platform,
    string DeviceName,
    string AppVersion,
    long FirstSeenAt,
    long LastSeenAt,
    long? RevokedAt,
    bool Current);

internal sealed record AuthTokens(
    string AccessToken,
    long AccessExpiresAt,
    string RefreshToken,
    long RefreshExpiresAt,
    string OfflineLicense,
    long OfflineLicenseExpiresAt,
    long ServerTime);

internal sealed record StoredSession(UserProfile User, AuthTokens Tokens);

internal sealed record StoredAuthState(
    string InstallationId,
    StoredSession? Session,
    long? LastTrustedServerTime);

internal sealed record BackendAuthorization(
    string OfflineLicense,
    string InstallationHash,
    string PublicJwk,
    string OmicsKeyB64);

internal sealed record AuthConfiguration(string BaseUrl, string PublicJwk)
{
    public static AuthConfiguration Load()
    {
        var overrideUrl = Environment.GetEnvironmentVariable("MY_BIO_TOOLS_AUTH_BASE_URL");
        var overrideKey = Environment.GetEnvironmentVariable("MY_BIO_TOOLS_LICENSE_PUBLIC_JWK");
        if (!string.IsNullOrWhiteSpace(overrideUrl) && !string.IsNullOrWhiteSpace(overrideKey))
        {
            return Validate(new AuthConfiguration(overrideUrl, overrideKey));
        }

        var path = Path.Combine(AppContext.BaseDirectory, "auth-config.json");
        if (!File.Exists(path))
        {
            throw new AuthConfigurationException("安装包缺少 auth-config.json 授权配置。");
        }
        var configuration = System.Text.Json.JsonSerializer.Deserialize<AuthConfiguration>(
            File.ReadAllText(path), AuthJson.Options)
            ?? throw new AuthConfigurationException("授权配置格式无效。");
        return Validate(configuration);
    }

    private static AuthConfiguration Validate(AuthConfiguration configuration)
    {
        if (!Uri.TryCreate(configuration.BaseUrl, UriKind.Absolute, out var uri)
            || uri.Scheme != Uri.UriSchemeHttps
            || string.IsNullOrWhiteSpace(configuration.PublicJwk))
        {
            throw new AuthConfigurationException("授权服务地址或生产公钥尚未配置。");
        }
        return configuration with { BaseUrl = configuration.BaseUrl.TrimEnd('/') };
    }
}

internal sealed class AuthConfigurationException(string message) : Exception(message);

internal sealed class AuthApiException(string code, string message, int statusCode) : Exception(message)
{
    public string Code { get; } = code;
    public int StatusCode { get; } = statusCode;
    public bool IsExplicitRevocation => statusCode is 401 or 403
        || Code is "AUTHORIZATION_REVOKED" or "AUTHORIZATION_EXPIRED" or "ACCOUNT_SUSPENDED" or "ACCOUNT_DELETED" or "SESSION_EXPIRED";
}

internal sealed record OfflineLicenseClaims(
    [property: JsonPropertyName("typ")] string Type,
    [property: JsonPropertyName("sub")] string Subject,
    [property: JsonPropertyName("device")] string Device,
    [property: JsonPropertyName("iat")] long IssuedAt,
    [property: JsonPropertyName("exp")] long ExpiresAt,
    [property: JsonPropertyName("version")] int Version,
    [property: JsonPropertyName("omics_key_b64")] string OmicsKeyB64);

internal static class AuthJson
{
    public static readonly System.Text.Json.JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.CamelCase,
    };
}
